qiskit/compiler/transpiler.py

```diff
@@ -634,28 +648,39 @@ def _parse_inst_map(inst_map, backend, num_circuits):
-                if faulty_map:
-                    faulty_edges = [gate.qubits for gate in backend.properties().faulty_gates()]
-                    functional_gates = [
-                        edge for edge in configuration.coupling_map if edge not in faulty_edges
-                    ]
-                    coupling_map = CouplingMap()
-                    for qubit1, qubit2 in functional_gates:
-                        if faulty_map[qubit1] is not None and faulty_map[qubit2] is not None:
-                            coupling_map.add_edge(faulty_map[qubit1], faulty_map[qubit2])
-                    if configuration.n_qubits != coupling_map.size():
-                        warnings.warn(
-                            "The backend has currently some qubits/edges out of service."
-                            " This temporarily reduces the backend size from "
-                            f"{configuration.n_qubits} to {coupling_map.size()}",
-                            UserWarning,
-                        )
-                else:
-                    coupling_map = CouplingMap(configuration.coupling_map)
+        backend_version = getattr(backend, "version", 0)
+        if backend_version <= 1:
+            if getattr(backend, "configuration", None):
+                configuration = backend.configuration()
+                if hasattr(configuration, "coupling_map") and configuration.coupling_map:
+                    faulty_map = _create_faulty_qubits_map(backend)
+                    if faulty_map:
+                        faulty_edges = [gate.qubits for gate in backend.properties().faulty_gates()]
+                        functional_gates = [
+                            edge for edge in configuration.coupling_map if edge not in faulty_edges
+                        ]
+                        coupling_map = CouplingMap()
+                        for qubit1, qubit2 in functional_gates:
+                            if faulty_map[qubit1] is not None and faulty_map[qubit2] is not None:
+                                coupling_map.add_edge(faulty_map[qubit1], faulty_map[qubit2])
+                        if configuration.n_qubits != coupling_map.size():
+                            warnings.warn(
+                                "The backend has currently some qubits/edges out of service."
+                                " This temporarily reduces the backend size from "
+                                f"{configuration.n_qubits} to {coupling_map.size()}",
+                                UserWarning,
+                            )
+                    else:
+                        coupling_map = CouplingMap(configuration.coupling_map)
+
+        else:
+            coupling_map = backend.coupling_map
+            faulty_map = _create_faulty_qubits_map(backend)
+            if faulty_map:
+                coupling_map = CouplingMap()
+                for qubit1, qubit2 in configuration.coupling_map:
```

Comment:
This code still needs to be updated (and is why tests are failing). I agree this isn't correct as is now and I'll have to update this before I remove the WIP. For this PR I'm going to just update `transpile()` to support the same model it does currently with `BackendV1` (which is why there are `coupling_map()` and `duration()` methods to get the existing transpiler data structures from a `Target`).

After this PR merges I will work on a transpiler PR to start taking advantage of the extra information that a BackendV2 provides. For some passes this will be a lot of work (like routing passes for example). I'm thinking the first pass will just be making the basis translator qargs aware so after we map and route the basis translator will know which gates are available on which qubits (or on which sets of qubits) and use that when making translation decisions.

For this chunk of code in particular I haven't though too much about how the "Faulty Qubits" model works with BackendV2. Honestly I feel like this is something the qiskit-ibm provider should be handling before terra sees it because the ibm backends are the only thing which could report this.

Reply:
I've updated the `transpile()` code in: https://github.com/Qiskit/qiskit-terra/pull/5885/commits/4a76c0be68d3e41ac5fe211eb686e906ecd69917 to start to correctly use the BackendV2 api. I still need to add support for things from the former `backend.properties()` attributes and tests that actually use BackendV2 though